### PR TITLE
add remat docstring

### PR DESCRIPTION
**Pull request by mattjj**
first part of addressing #3314

### CODE REVIEW

File: jax/api.py [hunk 1]
```diff
@@ -1788,6 +1788,85 @@ def abstractify(x):
...
+  reused on the backward pass. This evaluation strategy can lead to a high
+  memory cost, or even to poor performance on hardware acceleartors where memory
+  access is much more expensive than FLOPs.
+
+  An alternative evaluation strategy is for some of the linearization points to
+  be recomputed (i.e. rematerialized) rather than stored. This approach can
+  reduce memory usage at the cost of increased computation.
+
+  This function decorator produces a new version of ``fun`` which follows the
+  rematerialization strategy rather than the default store-eveything strategy.
```
Comment (hawkinsp) [inline jax/api.py:1810]: sp everything

File: jax/api.py [hunk 2]
```diff
@@ -1788,6 +1788,85 @@ def abstractify(x):
...
+      scalars, or (nested) standard Python containers (tuple/list/dict) thereof.
+    concrete: Optional, boolean indicating whether ``fun`` may involve
+      value-dependent Python control flow (default False). Support for such
+      control flow is optional, and disabled by default, because in some
+      edge-case compositions with ``jax.jit`` it can lead to some extra
+      computation.
+
+  Returns:
+    A function (callable) with the same input/output behavior as ``fun`` but
+    which, when differentiated using e.g. ``jax.grad``, ``jax.vjp``, or
```
Comment (hawkinsp) [inline jax/api.py:1830]:
```
:func:`jax.grad`
```
etc.

File: jax/api.py [hunk 3]
```diff
@@ -1788,6 +1788,85 @@ def abstractify(x):
 
 
 def checkpoint(fun: Callable, concrete: bool = False) -> Callable:
+  """Make ``fun`` recompute internal linearization points when differentiated.
+
+  The ``jax.checkpoint`` decorator, aliased to ``jax.remat``, provides a way to
+  trade off computation time and memory cost in the context of automatic
+  differentiation, especially with reverse-mode autodiff like ``jax.grad`` and
```
Comment (hawkinsp) [inline jax/api.py:1795]:
```
:func:`jax.grad`
```
and similarly everywhere.

**Review by hawkinsp — APPROVED** [review]

File: jax/api.py (hunk 1 above)
Comment (mattjj) [reply jax/api.py]: Nice catch!

**Comment by mattjj** [comment]
@tomhennigan I'm still keen on your feedback, but let's land and iterate!

**Comment by shoyer** [comment]
It would probably also make sense to include this in the API docs, too?

**Comment by mattjj** [comment]
@shoyer You are right, thanks for catching that! An oversight on my part.

**Pull request merged**